libkbfs/prefetcher.go

```diff
@@ -31,22 +40,42 @@ type blockRetriever interface {
 }
 
 type blockPrefetcher struct {
-	retriever  blockRetriever
+	config prefetcherConfig
+	log    logger.Logger
+	// blockRetriever to retrieve blocks from the server
+	retriever blockRetriever
+	// channel to synchronize prefetch requests with the prefetcher shutdown
 	progressCh chan prefetchRequest
+	// channel that is idempotently closed when a shutdown occurs
 	shutdownCh chan struct{}
-	doneCh     chan struct{}
+	// channel that is closed when a shutdown completes and all pending
+	// prefetch requests are complete
+	doneCh chan struct{}
 }
 
 var _ Prefetcher = (*blockPrefetcher)(nil)
 
-func newBlockPrefetcher(retriever blockRetriever) *blockPrefetcher {
+func newBlockPrefetcher(retriever blockRetriever, config prefetcherConfig) *blockPrefetcher {
 	p := &blockPrefetcher{
+		config:     config,
 		retriever:  retriever,
 		progressCh: make(chan prefetchRequest),
 		shutdownCh: make(chan struct{}),
 		doneCh:     make(chan struct{}),
 	}
-	go p.run()
+	if config != nil {
```

Comment:
ah, i see it's for returning a nil one for mocks. Hmm, I think having an `emptyPrefetcher` implementation of `Prefetcher` is cleaner, but I don't feel too strongly about it